### PR TITLE
Prevent running again already running cron group

### PR DESCRIPTION
**Pull request by paveq**
### Description
Prevents running the same cron group concurrently. Alternative implementation to https://github.com/magento/magento2/pull/11465

### Fixed Issues (if relevant)
1. magento/magento2#10650: Cron starts when it's already running

### Manual testing scenarios
1. Define cron job that lasts a while, or add _sleep()_ call under __runJob()_ method
2. Run cron:run entry point multiple times

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by paveq** [comment]
@dmanners @ihor-sviziev I've created this MVP implementation of cron group based locking. This is now being tested in Vaimo internally too.

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Lock/Backend/Database.php]: As magento 2.2 and later support php 7+ only, could you scecify types for these variables? Same for other methods in this file + interface.

Also would be great to add declare strict types for new file.

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Lock/Backend/Database.php]: Could you import this resource connection, it will be more readable there.

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Lock/Backend/Database.php]: Would be great to cover this class with unit and integration tests. Could you add them?

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/Lock/Backend/Database.php]: Another thing:
> MySQL 5.7.5 and later enforces a maximum length on lock names of 64 characters. Previously, no limit was enforced.

I think it should be checked in order to prevent mysql errors there

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @paveq,
Thank you so much for such good PR! 
I think there should be done few fixes. Could you review my comments?

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (paveq) [reply lib/internal/Magento/Framework/Lock/Backend/Database.php]: Sure, I'll work on implementing some test cases for this.

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (paveq) [reply lib/internal/Magento/Framework/Lock/Backend/Database.php]: Done.

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (paveq) [reply lib/internal/Magento/Framework/Lock/Backend/Database.php]: Implemented scalar type hints.

**Comment by paveq** [comment]
In order for locks not to conflict in setups where multiple Magento installations are running in single database, we need to use some kind of installation specific unique prefix for the lock.

**Framework\Cache\Core::_id()** and **Framework\App\Cache\Frontend\Factory** already contain similar mechanism that is used by cache, and could be used by locks too. However that method is protected, and injecting cache core to this code seems not a correct solution anyway.

Would it make sense for this unique installation specific id generation to be moved away from Cache Core, and be available in a generic way? (I'm suggesting that in scope of this PR we make such mechanism/API interface available, but not refactor Cache Core yet, otherwise the scope might grow to be quite big).

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (paveq) [reply lib/internal/Magento/Framework/Lock/Backend/Database.php]: Test cases added.

**Comment by paveq** [comment]
> Before MySQL 5.7.5, only a single simultaneous lock (per session) can be acquired and GET_LOCK() releases any existing lock (on the same connection).
>
> In MySQL 5.7.5, GET_LOCK() was reimplemented using the metadata locking (MDL) subsystem and its capabilities were extended. Multiple simultaneous locks can be acquired and GET_LOCK() does not release any existing locks. 

Does this pose an issue? In Magento system requirements we support MySQL 5.6 onwards.

Until system requirements have been raised to MySQL 5.7.5, perhaps we can check that only one lock can be acquired through LockManager at once? Ideally we would support multiple locks from MySQL 5.7.5 upwards, but at this moment that feature is not needed (YAGNI), and there appears to be no easy way to detect MySQL version from the connection.

File: dev/tests/integration/testsuite/Magento/Framework/Lock/Backend/DatabaseTest.php
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Framework/Lock/Backend/DatabaseTest.php]: Could you use private visibility instead of protected (for all methods in this file)?

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Lock/Backend/Database.php]: Could you add ```@throws``` to phpdoc block there?

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Lock/Backend/Database.php]: Could you add phpdoc block there?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi,

I don't think we need to check mysql version there. Current implementation looks good.

One thing: now Magento\Cron\Test\Unit\Observer\ProcessCronQueueObserverTest fails, could you update it?

**Review by ihor-sviziev — APPROVED** [review]
Looks really good!

**Comment by paveq** [comment]
Can you please comment on https://github.com/magento/magento2/pull/12497#issuecomment-348412843 ?

**Comment by ihor-sviziev** [comment]
@paveq good idea, but cache prefix might be really long. I suggest to use mysql database name as prefix as it is described in mysql implementation. I don't think that someone will try to use two magento installations in one mysql DB.

>GET_LOCK() can be used to implement application locks or to simulate record locks. Names are locked on a server-wide basis. If a name has been locked within one session, GET_LOCK() blocks any request by another session for a lock with the same name. This enables clients that agree on a given lock name to use the name to perform cooperative advisory locking. But be aware that it also enables a client that is not among the set of cooperating clients to lock a name, either inadvertently or deliberately, and thus prevent any of the cooperating clients from locking that name. One way to reduce the likelihood of this is to use lock names that are database-specific or application-specific. For example, use lock names of the form **db_name.str** or app_name.str. 

Also there should be ability to replace this prefix by plugins.

Actually we could use one of these options: mysql db name OR cache prefix. I prefer to have mysql db name

**Comment by paveq** [comment]
@ihor-sviziev I've implemented changes for using DB name as prefix, and left it overridable via DI.

**Comment by ihor-sviziev** [comment]
@paveq will not be able to process your PR till mid next week.
@magento-engcom-team could you review it?

**Comment by dmanners** [comment]
@paveq I will aim to get around to this PR in the next week. Sorry for the delay.

**Comment by ihor-sviziev** [comment]
From @diglin in https://github.com/magento/magento2/pull/11465#issuecomment-351672547:
> The solution provided by the other here raise an other issue: the whole cron group will be blocked if one cron job is running. If it last too long, the next run after the lock is released will won't necessary run the cron job because they are "missed" due to too late run. So voilà ! Problem not solved but my PR is closed

Need to check if we can eliminate it.

**Comment by paveq** [comment]
@ihor-sviziev Locking can be done in either job based or cron group based. this is a design and architecture decision Magento needs to take, for me as an individual contributor it's pretty hard to make this design decision.  If the locking would be done on job level, it raises question what is the point of cron groups then? It would be quite useful to use cron groups as logical group of tasks that somewhat relate to each other which need to run in sequence.

Cron executions can be missed even now, if one job takes too long and blocks entire default group (and thus cron scheduling). My PR does not really change that fact. It's responsibility of one who defines custom cron tasks/groups to have proper schedule times and cron group parameters (timeout for missed runs).

For some purposes it is perfectly fine to have missed jobs, let's take example: cronjob that runs every minute and takes items from queue, and it takes 5 minutes to process an item. Sometimes queue is empty, thus the job exists immediately. Despite processing if there's item taking 5 minutes we want to run every minute, so that we catch the item as soon as possible.

We can also define more cron groups than "index" and "default" we have (by default) now if this is an issue.

Please describe the roadmap how Cron functionality in Magento should be developed. I can think of:
1. Remove cron groups entirely, lock on job code basis
2. Further develop cron group based way (perhaps split default Magento tasks)
3. Keep as it is

EDIT: Ideally in my mind perhaps we should go away with cron groups, run everything in a separate process, and define dependencies/conflicts per job basis (maybe even "after" and "before" definitions, similar to totals calculation ordering).  However this requires much more work and also probably is such big change, that it needs to be introduced in major release (eg. 2.3.x).

**Comment by dmanners** [comment]
I can talk with @ishakhsuvarov during mageConf and see what we think is a good option here.

**Comment by dmanners** [comment]
After discussion with @vrann we will try to get some of the core architectural team involved here. I feel as this is a big change to to the cron is working we should make sure that everyone is happy with the final solution.

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (buskamuza) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php:64]: I would not hard-code timeout. Some processes can take more than 5min based on the amount of the processed data. The easiest solution would be to make it configurable (either in Admin Panel or in config.php). Better solution could be to adjust automatically (e.g., make each process output something and if there is no output consider it hung and release the lock), but it might be a more complex solution.

File: lib/internal/Magento/Framework/Lock/Backend/Database.php
Comment (buskamuza) [inline lib/internal/Magento/Framework/Lock/Backend/Database.php]: Please add full name (including prefix) to the exception message, so it's easy to understand what name caused the ussed.

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (buskamuza) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: Regarding the groups vs jobs discussion. Groups are just logical grouping for the user. There is no guarantee that different groups would not have dependent jobs. So locking mechanism should be implemented per job.

Dependent jobs is a good idea for the future cron evolution.

**Comment by buskamuza** [comment]
> Until system requirements have been raised to MySQL 5.7.5, perhaps we can check that only one lock can be acquired through LockManager at once?

Good idea. Please add this to the implementation.

**Review by buskamuza — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Lock/LockManagerInterface.php
Comment (kandy) [inline lib/internal/Magento/Framework/Lock/LockManagerInterface.php]: Maybe better to have lock($name), unlock($name), tryLock($name, $timeout) methods?

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (ihor-sviziev) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: Do we have any examples where one cron job has dependency to another one? Also currently they could be ran in parallel, so we already have this issue. Not sure we need to introduce fix for it as part of this PR.

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (paveq) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php:64]: This is actually five seconds, after that we give up and continue to the next group.

Reasoning being: *cron:run* is triggered from system crontab every minute, so next check will happen within so small period, that it does not make sense to wait for a long time. I suppose this could be reduced to even one or zero seconds (skip group immediately if lock could not be taken).

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (paveq) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: I previously commented that I would prefer cron groups being removed, but let's wait comments from Core team.

At this moment Groups are logical grouping, which would itself be ideal for this use. I can see scenario where example product import related jobs could be put into the same group, so that they do not run parallel, causing less stress on indexing and avoiding possible deadlocks.

**Comment by paveq** [comment]
>> Until system requirements have been raised to MySQL 5.7.5, perhaps we can check that only one lock can be acquired through LockManager at once?
> 
> Good idea. Please add this to the implementation.

I would prefer raising MySQL system requirement as solution, if this turns to be bigger PR that will be merged to major release 2.3.

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (buskamuza) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php:64]: Makes sense. Thanks. I think it's ok to leave it as is (5s).

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (buskamuza) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: For now, groups are used for grouping settings (cron run frequency) and it's not guaranteed that jobs from different groups can be run in parallel. As I mentioned before, I agree that it would be good to implement cron job dependencies, so some jobs could run in parallel. For now, jobs should run consequently, independently on the group.

Note: I represent core team, and discussed this question internally.

**Comment by buskamuza** [comment]
> I would prefer raising MySQL system requirement as solution, if this turns to be bigger PR that will be merged to major release 2.3.

We're reviewing versions of 3rd-party libraries and tools right now. Don't have an answer to your request yet. So, at the moment we should assume that Magento is going to support MySQL 5.6.

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (paveq) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: @buskamuza I am not sure how to proceed with this PR after your comment. We can move locking mechanism outside "foreach" loop and have single global lock for cron runner, at least as a stop gap measure.

Yet I would like to see Magento being able to run at least some tasks parallel. Cron groups currently implement a toggle "run in a separate process". Would it be safe to assume those groups can run parallel to main process, otherwise what would be the purpose of this setting?

Comment (buskamuza) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: As cron already runs per groups, let's just leave your implementation of locking in scope of this PR.
In future, we may reconsider it and reimplement it based on job dependencies.

Comment (buskamuza) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: Just for clarification: no further actions are needed here. Just review other comments in the PR. Thanks.

**Comment by ihor-sviziev** [comment]
@paveq as we discussed with @buskamuza - this PR can be accepted after all requested changes (except cron groups locks vs cron job locks). Could you update your PR?

File: lib/internal/Magento/Framework/Lock/LockManagerInterface.php
Comment (paveq) [reply lib/internal/Magento/Framework/Lock/LockManagerInterface.php]: I can change to *lock* and *unlock* if there's consensus these are better action verbs. However I would avoid *tryLock* as ambiguous term, for example in Java is has different meaning than just testing the lock status. However *testLock* could be one alternative?

EDIT: *lock* and *unlock* seem to be common terms in eg. Linux kernel and also Java, changing this implementation to conform to those wise spread terms.

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (paveq) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: After internal discussion with a development team that faced some issue with 3rd party module, we thought that maybe bit more strict path could be taken here. Implementation could be changed to use single global lock for all groups **except** for those that have "*run in a separate process*" enabled.

Thoughts on this? I would say it's safer way to go. However even my current solution is better than what we have now, as there's no limitation whatsoever.

**Comment by paveq** [comment]
@ihor-sviziev @buskamuza Could you review the changes now?

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[MAGETWO-88237](https://jira.corp.magento.com/browse/MAGETWO-88237) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@paveq thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Review by buskamuza — APPROVED** [review]

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=12497) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=12497) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: paveq<br/>:x: kandy

**Comment by erikhansen** [comment]
@kandy I see that you added some commits to this PR recently. Do you anticipate there will be any additional changes to this PR before it's merged into the `2.2-develop` branch? I'm asking because I'm experiencing the issue described by #10650 on a production site and I'm considering applying this PR as a patch to fix this issue.

@paveq [You mentioned](https://github.com/magento/magento2/pull/12497#issuecomment-348200726) that "This is now being tested in Vaimo internally too". Are you running these changes in a production environment?

**Comment by kandy** [comment]
hi @erikhansen, currently this ticket in QA process, but I don't expect  additional commits

**Comment by paveq** [comment]
@erikhansen Yes we are running the changes in production.

**Comment by sidolov** [comment]
Fixed static failures 
@erikhansen I think it was last changes for this PR

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php
Comment (adrian-martinez-interactiv4) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php:174]: Shouldn't BC be preserved?

Comment (kandy) [reply app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php:174]: @adrian-martinez-interactiv4, We treat Observers and Plugins as internal implementation and don't try to preserve BC for them

**Comment by erikhansen** [comment]
I've applied this patch to two production sites. While it does seem to resolve the duplicate CRON issue, it seems to have introduced a regression: The `cron_schedule` table does not seem to be getting cleaned according to the "History Lifetime" settings. I have set the lifetime value at 180 (3 hours), however there are many records in the `cron_schedule` table from the previous couple of weeks.

Has anyone else seen this issue? I didn't open a separate Github issue for this, as this code isn't in a GA release yet, so I can't imagine many have run into this. In the https://github.com/magento/magento2/issues/11002#issuecomment-360581025 issue, someone suggested a solution to solve this problem, however it seems like that should not be necessary if the success lifetime is set to 180 for all CRON groups.

**Comment by kandy** [comment]
Hi @erikhansen. 

Looks like problem introduced by this code https://github.com/paveq/magento2/blob/175229e23b5df6473bebf7acd24356a38ca6c9e3/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L502

Need change it to 
```
foreach ($historyLifetimes as $status => $time) {
            $count += $connection->delete(
                $scheduleResource->getMainTable(),
                [
                    'status = ?' => $status,
                    'job_code in (?)' => array_keys($jobs),
                    'created_at < ?' => $connection->formatDate($currentTime - $time)
                ]
            );
        } 
```

Can I ask you to create separate issue for this?

**Comment by Ian410** [comment]
I implemented this on a 2.2.4 production site.  It definitely prevents duplicates from running but in our case, it didn't run jobs at all if one of the jobs ended up taking a long time.  Since it's locking groups I think if a job in the default group takes a long time it won't schedule or launch other groups; at least that's my working theory right now.  I had to revert this set of changes.

**Comment by ihor-sviziev** [comment]
@kandy  feel free to create PR for this, even without issue reporting

**Comment by erikhansen** [comment]
@kandy Can you expound on [the change you mentioned](https://github.com/magento/magento2/pull/12497#issuecomment-394324949) in your comment from a couple of weeks ago? The only difference I see in your codeblock is that this line:
```
'status = ?' => Schedule::STATUS_PENDING,
```

should be changed to:

```
'status = ?' => $status,
```

However if you look at [the source file](https://github.com/paveq/magento2/blob/175229e23b5df6473bebf7acd24356a38ca6c9e3/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L502), there is no `$status` variable defined. 

Can you let me know what this should be changed to, and I'll change it and test it on a stage environment? I can then create a PR for this, expounding on the issue along with your suggested fix.

**Comment by hostep** [comment]
@erikhansen, look at his updated `foreach` loop, you'll see the variable in there ;)
> `foreach ($historyLifetimes as $status => $time) {`

**Comment by erikhansen** [comment]
@hostep Thanks. I just realized that independently and was coming back here to post when I saw your comment. I'm deploying this to a stage environment that had 1.2M rows in `cron_schedule` and once I confirm this fixes the issue, I'll submit a PR to resolve this.

**Comment by georgebotley** [comment]
@erikhansen Can you confirm if this has fixed issues for you? We've got a few servers with production Magento sites on and it brings MySQL to its knees. It'd be great if this is a word around fixes things..

**Comment by erikhansen** [comment]
@torindul Applying the contents of this PR as a patch to a Magento 2.2.3 site fixes the issue with the same CRON job running at the same time. By the way, [this Gist](https://gist.github.com/erikhansen/d6edc492e1dbbb2dc58a0ccd794dd9a0) contains a query that will show you if you have a problem with this.

~However applying the patch introduced an issue where CRON jobs are not getting cleaned up, as I explained in [my comment from 27 days ago](https://github.com/magento/magento2/pull/12497#issuecomment-394112843). I made [the change that @kandy suggested](https://github.com/magento/magento2/pull/12497#issuecomment-394324949) but that did not solve my issue. So until I make time to investigate a proper solution for it, I've implemented a custom CRON job as mentioned [here](https://magento.stackexchange.com/questions/208592/magento-2-cronjob-bug-mysql-is-always-running-at-30-usage-and-many-php-proces/208597#208597).~ UPDATE: See [my comment from July 16th, 2018 below](https://github.com/magento/magento2/pull/12497#issuecomment-405428222).

For anyone wanting to apply this PR as a patch to their environment, here are patches that can be applied via [composer-patches](https://github.com/cweagans/composer-patches) (you'll need to remove `.txt` from the filenames). ~However be aware that is introduces a fairly serious regression of CRONs not being cleaned up.~ UPDATE: This was not actually accurate.

* [Patch to apply to `magento/magento2-base`](https://github.com/magento/magento2/files/2149967/MDVA-9408_EE_2.2.3_COMPOSER_v1-magento2-base.patch.txt)
* [Patch to apply to `magento/module-cron`](https://github.com/magento/magento2/files/2149968/MDVA-9408_EE_2.2.3_COMPOSER_v1-module-cron.patch.txt)
* [Patch to apply to `magento/framework`](https://github.com/magento/magento2/files/2149966/MDVA-9408_EE_2.2.3_COMPOSER_v1-framework.patch.txt)

**Comment by erikhansen** [comment]
I'm currently going through the process of upgrading a site from 2.2.3 > 2.2.5. In the process, I realized that the changes in this PR are included in 2.2.5. I expect that anyone upgrading to 2.2.5 will experience the issue with CRONs not getting cleaned up.

I will soon be upgrading a vanilla Magento site to 2.2.5 and if I'm able to reproduce the CRON issue, I'll open a Github issue with full details.

**Comment by kandy** [comment]
hi @erikhansen, thank for so deep investigation
please pay attention that fix for job cleanup already included in 2.2.5
https://github.com/magento/magento2/blob/2.2.5/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L506

**Comment by erikhansen** [comment]
@kandy Thanks. I was aware that the fix you referenced was included in 2.2.5. However I have confirmed that this is still an issue. I will be opening a new Github issue this morning and will post a link here once I've done that.

**Comment by kandy** [comment]
@erikhansen, Also, pay attention that cron history cleanup time was changed to 7 days in commit
https://github.com/magento/magento2/commit/ba63d94066e2ab612297a920c134f1da8e89d65a

**Comment by erikhansen** [comment]
@kandy I finally made time to look into this. I was wrong: 2.2.5 does _not_ have an issue with CRONs that never get deleted. What I was seeing was exactly what you pointed out: the `cron_schedule` gets filled with so many rows because the default is now for Magento to store 7 days of history. This means that a vanilla M2 install will have 10K+ rows in the `cron_schedule` table on a regular basis.

In 2.2.4 and prior, having a large row count in the `cron_schedule` table [occasionally caused issues](https://github.com/magento/magento2/issues/11002#issuecomment-376675806) since the entries in that table were deleted one-by-one—that is why [I was initially concerned](https://github.com/magento/magento2/pull/12497#issuecomment-401410917) about 2.2.5. However now that the `DELETE` query has been optimized, this should no longer be an issue.


For the record, here's what a `DELETE` statement looks like on 2.2.5:
```
DELETE FROM `cron_schedule` 
    WHERE (status = 'pending') 
        AND (job_code in ('staging_apply_version', 'staging_remove_updates', 'staging_synchronize_entities_period')) 
        AND (created_at < '2018-05-31 17:55:04')
```

Compared to 2.2.4 and before:
```
DELETE FROM `cron_schedule` WHERE (schedule_id = '6594032')
```

**Comment by cscd98** [comment]
@paveq @hostep @sidolov @magento-engcom-team @nmalevanec 

I am getting:

```
[2019-07-08 16:11:08] main.WARNING: Could not acquire lock for cron group: index, skipping run [] []
[2019-07-08 16:11:08] main.WARNING: Could not acquire lock for cron group: ddg_automation, skipping run [] []
[2019-07-08 16:11:08] main.WARNING: Could not acquire lock for cron group: consumers, skipping run [] []

```

I have emptied cron_schedule, still same issue. How do I clear the locks down???

This is how my cron runs:

```
flock -w 1 /var/www/vhosts/xx/current/var/cron.lock php -f /var/www/vhosts/xx/current/bin/magento cron:run | grep -v "Ran jobs by schedule" >> /var/www/vhosts/xx/current/var/log/magento.cron.log
```

**Comment by kandy** [comment]
@craigcarnell Can I ask you create separate issue for this? Also, can you describe your environment?

**Comment by hostep** [comment]
@craigcarnell: it might have something to do with your `cron.lock` file existing in the `current` directory. Which means you could potentially have 2 or more cron executions still running next to each other just after you deployed a new release. I'd advise you to put the `cron.lock` file in a directory which doesn't change in between different deploys.
Also: we are no longer use the `flock` "fix", Magento 2.2.5 and later now handle locking on the database level since this PR got merged.

**Comment by mowny** [comment]
We're still running 2.2.8. It doesn't run very fast, though. A big part of this was cronjobs piling up due to taking longer than a minute to run because of mysql congestion.

Examining cron_schedule, I found:
`291:        if ($scheduledTime < $currentTime - $scheduleLifetime) {`
should move planned entries from pending to missed after "Missed if not Run Within" expired, correct? Then why do I have thousands of entries in cron_schedule that are still pending, and only being purged after "Failure History Lifetime"? (The oldest I see match the 4320 minutes "Failure History Lifetime".)

IMO there should ever only be _one_ scheduled job of a given type in "pending" state which is past its scheduled time, as all others that were scheduled even earlier can be considered "missed"; there is no benefit in having them still pending, as multiple runs will not "catch up" any better than a single run.

**Pull request merged**
